IndexedDB should properly handle names containing unpaired UTF-16 surrogates

szewai opened the pull request:
#### 4905ec5f9240c6826dca52cfea2ae496fbb7b8b3
<pre>
IndexedDB should properly handle names containing unpaired UTF-16 surrogates
<a href="https://bugs.webkit.org/show_bug.cgi?id=300902">https://bugs.webkit.org/show_bug.cgi?id=300902</a>
<a href="https://rdar.apple.com/161082712">rdar://161082712</a>

Reviewed by Chris Dumez.

SQLiteIDBBackingStore currently uses UTF-8 encoding for database name. UTF-8 is designed to encode valid Unicode code
points, and invalid code points like lone UTF-16 surrogate (e.g. U+D9B2) will be replaced with special bytes. According
to spec, database name is DOMString, which can contain unpaired UTF-16 surrogates. This means the database name stored
in database (with UTF-8 encoding) can be different from database name created from JavaScript.

To preserve the exact UTF-16 data, now we make SQLiteIDBBackingStore store database name as blob instead of text. As
SQLite has dynamic typing, we don&apos;t need to change schema for this. For existing databases, we upgrade its metadata
version and database name on open, if the stored UTF-8 encoded name matches UTF-8 encoded requested name.

The patch also includes a change to close IndexedDB SQLite database connection when it becomes idle (no more client and
no in-memory data). This would help avoid resource leakage (fd, memory, etc), and it allows us to test re-open behavior
(with IDBDatabase.close) in the new tests.

Tests: storage/indexeddb/database-name-with-surrogates-private.html
       storage/indexeddb/database-name-with-surrogates.html

* LayoutTests/storage/indexeddb/database-name-with-surrogates-expected.txt: Added.
* LayoutTests/storage/indexeddb/database-name-with-surrogates-private-expected.txt: Added.
* LayoutTests/storage/indexeddb/database-name-with-surrogates-private.html: Added.
* LayoutTests/storage/indexeddb/database-name-with-surrogates.html: Added.
* LayoutTests/storage/indexeddb/resources/database-name-with-surrogates.js: Added.
(test.request2.onsuccess):
(test.request.onsuccess):
(test):
(testIDBFactoryDatabases):
* Source/WebCore/Modules/indexeddb/server/SQLiteIDBBackingStore.cpp:
(WebCore::IDBServer::SQLiteIDBBackingStore::createAndPopulateInitialDatabaseInfo):
(WebCore::IDBServer::getDatabaseNameFromDatabase):
(WebCore::IDBServer::migrateIDBDatabaseInfoTableIfNecessary):
(WebCore::IDBServer::SQLiteIDBBackingStore::extractExistingDatabaseInfo):
(WebCore::IDBServer::SQLiteIDBBackingStore::databaseNameAndVersionFromFile):
* Source/WebKit/NetworkProcess/storage/IDBStorageManager.cpp:
(WebKit::IDBStorageManager::tryCloseDatabase):
* Source/WebKit/NetworkProcess/storage/IDBStorageManager.h:
* Source/WebKit/NetworkProcess/storage/NetworkStorageManager.cpp:
(WebKit::NetworkStorageManager::databaseConnectionClosed):

Canonical link: <a href="https://commits.webkit.org/301753@main">https://commits.webkit.org/301753@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/19b6cdda575d184e361c10782bc5fe120b3fd77f

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/126915 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/46554 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/37540 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/133919 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/78500 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/b32e33f3-8940-4626-ae28-62ebe78b11ef) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/128786 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/47175 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/55085 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/96578 "Passed tests") | [❌ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/64582 "") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/24f44344-6a0b-4a52-8876-55e791e60f69) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/129863 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/37752 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/113524 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/77092 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/debc9524-14e8-4c0f-b655-2a441b603781) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/36615 "Passed tests") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/135/builds/31692 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/77313 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/107586 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/32008 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/136444 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/53578 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/41250 "Passed tests") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/105096 "") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/54080 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/109885 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/104788 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/26724 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/50301 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/28639 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/51056 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/53509 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/59376 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/52753 "Built successfully") | | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/56087 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/54509 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WebKit/NetworkProcess/storage/IDBStorageManager.cpp:
```diff
@@ -355,4 +356,16 @@ void IDBStorageManager::handleLowMemoryWarning()
+
+    if (iterator->value->tryClose()) {
+        RELEASE_LOG(IndexedDB, "%p - IDBStorageManager::tryCloseDatabase: Closed database", this);
+        m_databases.remove(iterator);
```
This code looks a little dangerous. If `tryClose()` were to modify `m_databases` then it would invalidate your iterator and you'd have a security bug on this line. Given that this is not hot code, I don't see a reason to take risks and I would use `get(identifier)`  above and `remove(identifier)` here.

cdumez commented on Source/WebCore/Modules/indexeddb/server/SQLiteIDBBackingStore.cpp:
```diff
@@ -65,8 +65,11 @@ constexpr auto v2ObjectStoreInfoSchema = "CREATE TABLE ObjectStoreInfo (id INTEG
+ * Version 1 - Initial version.
+ * Version 2 - Store database name as blob.
+ */
+static const int currentMetadataVersion = 2;
```
`constexpr`

szewai committed:
IndexedDB should properly handle names containing unpaired UTF-16 surrogates
https://bugs.webkit.org/show_bug.cgi?id=300902
rdar://161082712

Reviewed by Chris Dumez.

SQLiteIDBBackingStore currently uses UTF-8 encoding for database name. UTF-8 is designed to encode valid Unicode code
points, and invalid code points like lone UTF-16 surrogate (e.g. U+D9B2) will be replaced with special bytes. According
to spec, database name is DOMString, which can contain unpaired UTF-16 surrogates. This means the database name stored
in database (with UTF-8 encoding) can be different from database name created from JavaScript.

To preserve the exact UTF-16 data, now we make SQLiteIDBBackingStore store database name as blob instead of text. As
SQLite has dynamic typing, we don't need to change schema for this. For existing databases, we upgrade its metadata
version and database name on open, if the stored UTF-8 encoded name matches UTF-8 encoded requested name.

The patch also includes a change to close IndexedDB SQLite database connection when it becomes idle (no more client and
no in-memory data). This would help avoid resource leakage (fd, memory, etc), and it allows us to test re-open behavior
(with IDBDatabase.close) in the new tests.

Tests: storage/indexeddb/database-name-with-surrogates-private.html
       storage/indexeddb/database-name-with-surrogates.html

* LayoutTests/storage/indexeddb/database-name-with-surrogates-expected.txt: Added.
* LayoutTests/storage/indexeddb/database-name-with-surrogates-private-expected.txt: Added.
* LayoutTests/storage/indexeddb/database-name-with-surrogates-private.html: Added.
* LayoutTests/storage/indexeddb/database-name-with-surrogates.html: Added.
* LayoutTests/storage/indexeddb/resources/database-name-with-surrogates.js: Added.
(test.request2.onsuccess):
(test.request.onsuccess):
(test):
(testIDBFactoryDatabases):
* Source/WebCore/Modules/indexeddb/server/SQLiteIDBBackingStore.cpp:
(WebCore::IDBServer::SQLiteIDBBackingStore::createAndPopulateInitialDatabaseInfo):
(WebCore::IDBServer::getDatabaseNameFromDatabase):
(WebCore::IDBServer::migrateIDBDatabaseInfoTableIfNecessary):
(WebCore::IDBServer::SQLiteIDBBackingStore::extractExistingDatabaseInfo):
(WebCore::IDBServer::SQLiteIDBBackingStore::databaseNameAndVersionFromFile):
* Source/WebKit/NetworkProcess/storage/IDBStorageManager.cpp:
(WebKit::IDBStorageManager::tryCloseDatabase):
* Source/WebKit/NetworkProcess/storage/IDBStorageManager.h:
* Source/WebKit/NetworkProcess/storage/NetworkStorageManager.cpp:
(WebKit::NetworkStorageManager::databaseConnectionClosed):

Canonical link: https://commits.webkit.org/301753@main
